File: samples/assistant/csharp-ooproc/AssistantSampleOutOfProc/TodoManager.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Threading.Tasks;
using Microsoft.Azure.Cosmos;
using Microsoft.Extensions.Logging;

namespace AssistantSample;

/// <summary>
/// Todo item record which gets stored in the database.
/// </summary>
/// <param name="Id">A unique ID for the todo item.</param>
/// <param name="Task">A description of the task.</param>
public record TodoItem(string Id, string Task);

/// <summary>
/// Interface for managing todo items.
/// </summary>
public interface ITodoManager
{
    /// <summary>
    /// Adds a new todo item to the database.
    /// </summary>
    Task AddTodoAsync(TodoItem todo);

    /// <summary>
    /// Gets all todo items from the database.
    /// </summary>
    Task<IReadOnlyList<TodoItem>> GetTodosAsync();
}

/// <summary>
/// Implementation of <see cref="ITodoManager"/> which stores todo items in memory.
/// </summary>
/// <remarks>
/// This implementation is designed to be used when running the function app locally.
/// </remarks>
class InMemoryTodoManager : ITodoManager
{
    readonly List<TodoItem> todos = new();

    public Task AddTodoAsync(TodoItem todo)
    {
        this.todos.Add(todo);
        return Task.CompletedTask;
    }

    public Task<IReadOnlyList<TodoItem>> GetTodosAsync()
    {
        return Task.FromResult<IReadOnlyList<TodoItem>>(this.todos.ToImmutableList());
    }
}

/// <summary>
/// Implementation of <see cref="ITodoManager"/> which stores todo items in Cosmos DB.
/// </summary>
/// <remarks>
/// This implementation can be used in production or when running the function app locally.
/// This implementation is also suitable for when the app needs to be scaled out or when
/// the todos need to be persisted even after the function app is recycled.
/// </remarks>
class CosmosDbTodoManager : ITodoManager
{
    readonly ILogger logger;
    readonly Container container;

    public CosmosDbTodoManager(ILoggerFactory loggerFactory, CosmosClient cosmosClient)
    {
        if (loggerFactory is null)
        {
            throw new ArgumentNullException(nameof(loggerFactory));
        }

        if (cosmosClient is null)
        {
            throw new ArgumentNullException(nameof(cosmosClient));
        }

        this.logger = loggerFactory.CreateLogger<CosmosDbTodoManager>();
        this.container = cosmosClient.GetContainer("testdb", "my-todos");
    }

    public async Task AddTodoAsync(TodoItem todo)
    {
        this.logger.LogInformation("Adding todo ID = {Id} to container '{Container}'.", todo.Id, this.container.Id);
        await this.container.CreateItemAsync(todo, new PartitionKey(todo.Id));
    }

    public async Task<IReadOnlyList<TodoItem>> GetTodosAsync()
    {
        this.logger.LogInformation("Getting all todos from container '{Container}'.", this.container.Id);
        FeedIterator<TodoItem> query = this.container.GetItemQueryIterator<TodoItem>(
            new QueryDefinition("SELECT * FROM c"));

        List<TodoItem> results = new();
        while (query.HasMoreResults)
        {
            FeedResponse<TodoItem> response = await query.ReadNextAsync();
            results.AddRange(response.Resource);
        }

        this.logger.LogInformation("Found {Count} todos in container '{Container}'.", results.Count, this.container.Id);
        return results;
    }
}

File: samples/assistant/java/src/main/java/com/azfs/TodoManager.java
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 */
package com.azfs;

import java.util.ArrayList;
import java.util.List;

import com.azure.cosmos.CosmosClient;
import com.azure.cosmos.CosmosContainer;
import com.azure.cosmos.CosmosDatabase;
import com.azure.cosmos.models.CosmosContainerResponse;
import com.azure.cosmos.models.CosmosDatabaseResponse;
import com.azure.cosmos.models.CosmosQueryRequestOptions;
import com.azure.cosmos.util.CosmosPagedIterable;

/**
 * Interface for managing todo items.
 */
public interface TodoManager {

    /**
     * Adds a new todo item.
     * @param todoItem The todo item to add.
     */
    public void addTodo(TodoItem todoItem);

    /**
     * Gets all todo items.
     * @return A list of all todo items.
     */
    public List<TodoItem> getTodos();

}

/**
 * Todo item pojo which gets stored in the database.
 */
class TodoItem {

    // A unique ID for the todo item.
    private String id;

    // A description of the task.
    private String task;

    public TodoItem(String id, String task) {
        this.id = id;
        this.task = task;
    }
}

/**
 * Implementation of {@link TodoManager} which stores todo items in memory.
 * This implementation is designed to be used when running the function app locally.
 * This implementation is also suitable for testing purposes.
 */
class InMemoryTodoManager implements TodoManager {

    private final List<TodoItem> todos = new ArrayList<>();

    @Override
    public void addTodo(TodoItem todoItem) {
        this.todos.add(todoItem);        
    }

    @Override
    public List<TodoItem> getTodos() {
        return this.todos;
    }

}

/**
 * Implementation of {@link TodoManager} which stores todo items in Cosmos DB.
 * This implementation can be used in production or when running the function app locally.
 * This implementation is also suitable for when the app needs to be scaled out or when
 * the todos need to be persisted even after the function app is recycled.
/ */
class CosmosDbTodoManager implements TodoManager {

    CosmosClient cosmosClient;
    CosmosDatabase database;
    CosmosContainer container;

    public CosmosDbTodoManager(CosmosClient cosmosClient) {

        if(cosmosClient == null)
            throw new IllegalArgumentException("cosmosClient is required");

        this.cosmosClient = cosmosClient;

        String databaseName = System.getenv("CosmosDatabaseName");
        String containerName = System.getenv("CosmosContainerName");

        if( databaseName == null || databaseName.isEmpty() || containerName == null || containerName.isEmpty())
            throw new IllegalArgumentException("CosmosDatabaseName and CosmosContainerName environment variables are required");

        CosmosDatabaseResponse cosmosDatabaseResponse =
                this.cosmosClient.createDatabaseIfNotExists(databaseName);
        this.database = this.cosmosClient.getDatabase(cosmosDatabaseResponse.getProperties().getId());

        CosmosContainerResponse cosmosContainerResponse =
                this.database.createContainerIfNotExists(containerName, "/id");
        this.container = database.getContainer(cosmosContainerResponse.getProperties().getId());
    }

    @Override
    public void addTodo(TodoItem todoItem) {
        this.container.createItem(todoItem);
    }

    @Override
    public List<TodoItem> getTodos() {
        CosmosQueryRequestOptions queryOptions = new CosmosQueryRequestOptions();
        CosmosPagedIterable<TodoItem> todoItemCosmosPagedIterable = this.container.queryItems("SELECT * FROM c", queryOptions, TodoItem.class);

        List<TodoItem> results = new ArrayList<>();

        todoItemCosmosPagedIterable.iterableByPage(10).forEach(cosmosItemPropertiesFeedResponse -> {
            results.addAll(cosmosItemPropertiesFeedResponse.getResults());
        });
        return results;
    }
        
}